Add support for unknown batch sizes in group_norm.

stefanwayon opened the pull request:
Previously, the `group_norm` layer did not work when the batch size was not known (as is the case when using e.g. `tf.data.Dataset.batch`). 

The change directly addresses this case by replacing a call to `Tensor.shape` with a call to `array_ops.shape`, as well as replacing `None` with `-1` when calling `array_ops.reshape`.

stefanwayon committed:
Add support for unknown batch sizes in group_norm.

Previously, the `group_norm` layer did not work when the batch size was not known (as is the case when using e.g. `tf.data.Dataset.batch`). 

The change directly addresses this case by replacing a call to `Tensor.shape` with a call to `array_ops.shape`, as well as replacing `None` with `-1` when calling `array_ops.reshape`.

SliMM commented:
I signed it...

SliMM commented:
@shlens

shlens commented:
Thank you for doing this. Please add a unit test to guarantee behavior is correct.

stefanwayon committed:
Add test for partially defined shapes in group_norm.

This test ensures that the `group_norm` op can handle unkown dimensions
that are neither along the reduction axes nor along the channel axis. It
creates a placeholder with a partially-defined shape, creates a
`group_norm` op, and then runs it feeding an array of ones instead of
the paceholder.

stefanwayon committed:
Switch to using `array_ops.shape` when computing `inputs_shape`.

Previously, `Tensor.shape` was used, and then `None`s were replaced with
`-1`. This only really works when there's only a single `None`, since
`inputs_shape` is used for reshaping a tensor. That would fail when
running the op if there are two or more `-1`s in the requested shapes.

SliMM commented:
Hey @shlens, I’ve now added the unit test, and actually changed the code to support an arbitrary number of unknown dimensions. Indexing into `TensorShape` is a bit ugly without slices, so I tried my best, but there might be better ways to do it.

stefanwayon committed:
Replace unnecessary calls to .

SliMM commented on tensorflow/contrib/layers/python/layers/normalization_test.py:
```diff
@@ -221,6 +221,18 @@ def testParamsShapeNotFullyDefinedChannelsAxis(self):
                                reduction_axes=[-3, -2])
 
+  def testParamsShapeNotFullyDefinedOtherAxes(self):
+    inputs = array_ops.placeholder(dtypes.float32, shape=(None, 3, 4, None, 32))
```
A good example of another axis that might not be fully defined is, perhaps, the sequence axis in RNN tensors. In that case you would have both the batch axis and the sequence axis potentially not fully defined.

I think it is important to have a test with at least two axes that are not fully defined. The implementation in the first commit of this patch would not pass such a test, whereas the latest one would.

Should I make a separate test for the batch axis, or should I rename this to something like `test ParamsShapeNotFullyDefinedTwoAxes`

tensorflowbutler commented:
Nagging Assignee @case540: It has been 18 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @case540: It has been 34 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @case540: It has been 49 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @case540: It has been 64 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @case540: It has been 79 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @case540: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

patzm commented:
any news on this issue? Or did anyone find a way to circumvent this issue by transforming the tensor before applying group norm? I tried `tf.map_fn(tensor, lambda x: group_norm(x, ...))` but without success.

stefanwayon committed:
Rename test for undefined batch and sequence axes.

stefanwayon committed:
Add test for undefined batch axis.

SliMM commented:
@patzm: If you use [`tf.data.Dataset`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) for your input pipeline, you can work around this issue by using either [`tf.contrib.data.batch_and_drop_remainder(batch_size)`](https://www.tensorflow.org/api_docs/python/tf/contrib/data/batch_and_drop_remainder), or [`tf.data.Dateset.batch(batch_size, drop_remainder=True)`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#batch), depending on your TensorFlow version. They both yield static batch sizes, but the last bit of your data will be dropped if necessary.

shlens reviewed (approved):
LGTM

martinwicke commented:
This changes contrib, which will disappear shortly. I'd prefer not to take this PR, and reopen a PR against github.com/tensorflow/addons, where these extensions should be hosted in future.
